drv/lpc55-sprot-server/src/main.rs

```diff
@@ -175,280 +114,618 @@ fn main() -> ! {
+// Container for spi and gpio
+struct Io {
+    spi: crate::spi_core::Spi,
+    gpio: drv_lpc55_gpio_api::Pins,
+    state: State,
+    stats: IoStats,
+}
+
+/// The state of the IO state machine
+///
+/// Logically, the sprot protocol operates in a request/response fashion,
+/// despite the fact that bytes must simultaneously be clocked in and out during SPI
+/// transactions.
+///
+/// As part of our logical state machine, the RoT can only be reading a request from
+/// the SP or writing a reply. There can only be one request in flight at a time.
+/// When desynchronization is detected it must be corrected.
+///
+/// The RoT always starts in State::Read(ReadState::WaitingForRequest);
+#[derive(Clone, Copy, PartialEq)]
+pub enum State {
+    Read(ReadState),
+    Write(WriteState),
+}
+
+impl State {
+    /// Initialize the state Io state machine
+    fn init() -> State {
+        State::Read(ReadState::WaitingForRequest)
+    }
+
+    /// Return the read state. Panics if the RoT is not currently reading.
+    fn read(&self) -> ReadState {
+        match self {
+            State::Read(s) => *s,
+            _ => panic!(),
         }
     }
+
+    /// Return the write state. Panics if the RoT is not currently writing.
+    fn write(&self) -> WriteState {
+        match self {
+            State::Write(s) => *s,
+            _ => panic!(),
+        }
+    }
+}
+
+/// The state of the IO state machine when trying to read a request from the SP
+#[derive(Clone, Copy, PartialEq)]
+pub enum ReadState {
+    /// The RoT is waiting for CSn to be asserted
+    WaitingForRequest,
+
+    /// The RoT is in frame and reading data until it sees CSn de-asserted
+    InFrame,
+
+    /// The frame was read cleanly.
+    ///
+    /// In this case we were:
+    ///   1. In `WaitingForRequest` and saw CSn asserted and CSn de-asserted at the
+    ///      same time with no overrun errors, and with data.
+    ///   2. In `InFrame` and saw CSn de-asserted with no overrun errors and
+    ///     with data.
+    ///
+    FrameRead,
+
+    /// The SP has pulsed CSn, which means that CSn was asserted and de-
+    /// asserted with no data clocked at all.
+    ///
+    /// The SP pulses CSn when it receives a ROT_IRQ it wasn't expecting. This
+    /// happens in the reply phase, and can occur if processing a request at
+    /// the RoT takes longer than a retry timeout at the SP.
+    ///
+    /// We can only transition to Flush in the `WaitingForRequest` or `InFrame`
+    /// states. Flush is a `transient` state indicating we should clear our
+    /// FIFOs and buffers and go back to `WaitingForRequest`.
+    Flush,
+
+    /// We missed reading some bytes. We should clear our buffers and fifos and
+    /// inform the SP.
+    Overrun,
+
+    /// We got an unexpected CSn Assert. This is either due to a weird timing
+    /// issue resulting in the next request being sent, or a CSn pulse.
+    ///
+    /// Either way we are desynchronized, and the safest thing to do is
```

Comment:
This comment can be cleaned up a bit to specify that the reply we are talking about sending is the one resulting from the current request being read in. There are now outstanding replies while reading requests or waiting for a CSn Assert to read a request in the current code.